### PR TITLE
UFP64

### LINKED ISSUE #37: UFP64
### Motivation

Addition of 64-bit fixed-point type - UFP64

### PR DESCRIPTION
**Pull request by rostyslavtyshko**
## Type of change

- New feature

## Changes

The following changes have been made:

Addition of 64-bit fixed-point typs - `UFP64`

## Notes

- Note 1

## Related Issues

Closes  #37

### CODE REVIEW

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 1]
```diff
@@ -0,0 +1,267 @@
+library ufp64;
+//! A wrapper around u64 type for a library for Sway for mathematical functions operating with signed 32.32-bit fixed point numbers.
+use core::num::*;
+use std::assert::assert;
+use std::math::*;
+use std::revert::revert;
+use std::u128::U128;
+use std::math::Exponentiate;
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:8]:
```suggestion
use std::{
     math::Exponentiate,
     u128::U128,
};
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+pub struct UFP64 {
+    value: u64,
+}
+
+impl UFP64 {
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:14]: Can you reorder these functions in order to follow the [contributing book guidelines](https://fuellabs.github.io/sway-applications/book/index.html)?

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    /// The smallest value that can be represented by this type.
+    pub fn min() -> Self {
+        Self {
+            value: ~u64::min(),
+        }
+    }
+
+    /// The largest value that can be represented by this type,
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:36]:
```suggestion
    /// The largest value that can be represented by this type.
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    pub fn denominator() -> u64 {
+        1 << 32
+    }
+
+    pub fn zero() -> Self {
+        Self { value: 0 }
+    }
+
+    /// Creates UFP64 from u64. Note that ~UFP64::from(1) is 1 / 2^32 and not 1.
+    pub fn from(value: u64) -> Self {
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:25]: Is this `from` used or should the `From<T>` from the standard library be used?

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+impl UFP64 {
+    /// Creates UFP64 that correponds to a unsigned integer
+    pub fn from_uint(uint: u64) -> Self {
+        Self {
+            value: ~Self::denominator() * uint,
+        }
+    }
+}
+
+impl UFP64 {
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:141]: There are multiple instances of `impl UFP64`. Could these be condensed into one for readability?

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+            // panic on overflow
+            revert(0);
+        }
+        Self {
+            value: nominator.lower,
+        }
+    }
+}
+
+trait Exponent {
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:244]: I think this should be exported to the standard library's math library and then imported into this file.

**Review by bitzoic — CHANGES_REQUESTED** [review]
This PR needs to be bumped to use `forc v0.30.1`. 

Are you also able to apply some of the comments made to the signed integers PR such as adding a [`mod.rs`](https://github.com/FuelLabs/sway-libs/blob/master/tests/src/test_projects/signed_i8/tests/mod.rs) for tests?

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 6 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:244]: Unfortunately no, due to this known and AFAIK yet not fixed issue

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 5 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:141]: Unfortunately no, due to this known and AFAIK yet not fixed issue

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 3 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw]: Fixed accordingly

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 1 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw]: Fixed accordingly

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 4 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw]: Fixed https://github.com/FuelLabs/sway-libs/pull/31/commits/a944d94d2c8498141b635066ff1429d8d93868fe

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 6 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw]: Fixed accordingly https://github.com/FuelLabs/sway-libs/pull/31/commits/035c90b7baadbacb7e189a91a5d932b6b115124f

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw (hunk 2 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw]: Tried my best to reorder in https://github.com/FuelLabs/sway-libs/pull/31/commits/4835449a514446eeb31959d7c76b16d25dc397e6 , but unfortunately due to known ordering issue it's not always possible

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    }
+}
+
+impl UFP64 {
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        64
+    }
+
+    /// Convinience function to know the denominator
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:27]:
```suggestion
    /// Convenience function to know the denominator.
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+        //coefficients in the Taylor series up to the seventh power
+        let p2 = UFP64::from(2147483648); // p2 == 1 / 2!
+        let p3 = UFP64::from(715827882); // p3 == 1 / 3!
+        let p4 = UFP64::from(178956970); // p4 == 1 / 4!
+        let p5 = UFP64::from(35791394); // p5 == 1 / 5!
+        let p6 = UFP64::from(5965232); // p6 == 1 / 6!
+        let p7 = UFP64::from(852176); // p7 == 1 / 7!
+        // common technique to counter loosing sugnifucant numbers in usual approximation
+        // Taylor series approximation of exponantiation function minus 1. The subtraction is done to deal with accuracy issues
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:237]:
```suggestion
        // Taylor series approximation of exponentiation function minus 1. The subtraction is done to deal with accuracy issues
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        let one = UFP64::from_uint(1);
+
+        //coefficients in the Taylor series up to the seventh power
+        let p2 = UFP64::from(2147483648); // p2 == 1 / 2!
+        let p3 = UFP64::from(715827882); // p3 == 1 / 3!
+        let p4 = UFP64::from(178956970); // p4 == 1 / 4!
+        let p5 = UFP64::from(35791394); // p5 == 1 / 5!
+        let p6 = UFP64::from(5965232); // p6 == 1 / 6!
+        let p7 = UFP64::from(852176); // p7 == 1 / 7!
+        // common technique to counter loosing sugnifucant numbers in usual approximation
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:236]:
```suggestion
        // Common technique to counter losing significant numbers in usual approximation
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        }
+    }
+}
+
+impl Exponent for UFP64 {
+    /// Exponent function. e ^ x
+    fn exp(exponent: Self) -> Self {
+        let one = UFP64::from_uint(1);
+
+        //coefficients in the Taylor series up to the seventh power
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:229]:
```suggestion
        // Coefficients in the Taylor series up to the seventh power
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+            return ceil;
+        }
+    }
+}
+
+impl Root for UFP64 {
+    /// Sqaure root for UFP64
+    fn sqrt(self) -> Self {
+        let nominator_root = self.value.sqrt();
+        // Need to multiple over 2 ^ 16, as the sqare root of the denominator
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:215]:
```suggestion
        // Need to multiply over 2 ^ 16, as the square root of the denominator 
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+impl UFP64 {
+    /// Returns the nearest integer to `self`. Round half-way cases away from
+    pub fn round(self) -> Self {
+        let floor = self.floor();
+        let ceil = self.ceil();
+        let diff_self_floor = self - floor;
+        let diff_ceil_self = ceil - self;
+
+        // check if we are nearer to the floor or to the ceiling
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:202]:
```suggestion
        // Check if we are closer to the floor or to the ceiling
```

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 13]
```diff
@@ -0,0 +1,264 @@
+library ufp64;
+//! A wrapper around u64 type for a library for Sway for mathematical functions operating with signed 32.32-bit fixed point numbers.
+use std::math::{Exponent, Exponentiate, Root};
+use std::u128::U128;
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:4]:
```suggestion
use std::{
     math::{Exponent, Exponentiate, Root},
     u128::U128
};
```
These can be combined

File: sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw [hunk 14]
```diff
@@ -0,0 +1,264 @@
+library ufp64;
+//! A wrapper around u64 type for a library for Sway for mathematical functions operating with signed 32.32-bit fixed point numbers.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw:2]:
```suggestion

// A wrapper library around the u64 type for mathematical functions operating with signed 64-bit fixed point numbers.

```

**Review by bitzoic — CHANGES_REQUESTED** [review]
Mostly just cleaning up some comments.

Are you able to add this to the test harness? The tests exist but none of them are called. You can follow this example of how the `i8` tests are run [here](https://github.com/FuelLabs/sway-libs/blob/master/tests/src/test_projects/signed_i8/tests/mod.rs). Here is the full PR that added all the tests: https://github.com/FuelLabs/sway-libs/pull/45

Are you also able to add a README and a SPECIFICATION so that other developers know how to use the library?

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+```
+let ufp64_169 = UFP64::from_uint(169);
+res = UFP64::sqrt(ufp64_169);
+```
+
+Power Function
+```
+let three = UFP64::from_uint(3);
+let five = UFP64::from_uint(5);
+res = five.pow(three);
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/README.md:55]:
```suggestion
let res = five.pow(three);
```
Not particularly important but it's nice to be able to copy and paste the examples and they just work.

File: sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md [hunk 16]
```diff
@@ -0,0 +1,53 @@
+# Overview
+
+This document provides an overview of the Unsigned Fixed Point 64 bit library.
+
+It outlines the use cases, i.e. specification, and describes how to implement the library.
+
+## Use Cases
+
+The Unsigned Fixed Point 64 bit can be used anytime a one needs fixed-point numbers.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md:9]:
```suggestion
The Unsigned Fixed Point 64 bit can be used anytime one needs fixed-point numbers.
```

or 

```suggestion
The Unsigned Fixed Point 64 bit can be used anytime a developer needs fixed-point numbers.
```

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
+# Overview
+
+The Unsigned Fixed Point 64 bit  library provides a library to use signed numbers in Sway. It has 1 distinct type: `UFP64`. This type is stack allocated.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/README.md:3]:
```suggestion
The Unsigned Fixed Point 64-bit library provides a library to use signed numbers in Sway. It has 1 distinct type: `UFP64`. This type is stack allocated.

```

File: sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
+# Overview
+
+This document provides an overview of the Unsigned Fixed Point 64 bit library.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md:3]:
```suggestion
This document provides an overview of the Unsigned Fixed Point 64-bit library.

```

**Review by bitzoic — CHANGES_REQUESTED** [review]
A few comments on the documentation but overall looks great!! This should be the last of the comments and it is ready for merge.

I also noticed it wasn't added to the [source README](https://github.com/FuelLabs/sway-libs/blob/master/README.md). Can you add it to the libraries list?

Still waiting on adding the library to the tests as described by the last review.

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+This type is stored as a `u64` under the hood. Therefore the size can be known at compile time and the length is static. 
+
+For more information please see the [specification](./SPECIFICATION.md).
+
+# Using the Library
+
+## Getting Started
+
+In order to use the `UFP64` type it must be added to the Forc.toml file and then imported into your Sway project. To add Sway-libs as a dependency to the Forc.toml in your project, please see the [README.md](../../../README.md).
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md (hunk 18 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md (hunk 17 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md (hunk 16 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/SPECIFICATION.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+Square Root function
+```
+let ufp64_169 = UFP64::from_uint(169);
+res = UFP64::sqrt(ufp64_169);
+```
+
+Power Function
+```
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+Exponential Function
+```
+let ten = UFP64::from_uint(10);
+res = UFP64::exp(ten);
+```
+
+Square Root function
+```
+let ufp64_169 = UFP64::from_uint(169);
+res = UFP64::sqrt(ufp64_169);
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+Mathematical functions
+
+Exponential Function
+```
+let ten = UFP64::from_uint(10);
+res = UFP64::exp(ten);
+```
+
+Square Root function
+```
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+// Subtract one signed value from another
+let ufp64_value_3 = ufp64_value_1 - ufp64_value_2;
+```
+
+Mathematical functions
+
+Exponential Function
+```
+let ten = UFP64::from_uint(10);
+res = UFP64::exp(ten);
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+let ufp64_value_3 = ufp64_value_1 + ufp64_value_2;
+
+// Subtract one signed value from another
+let ufp64_value_3 = ufp64_value_1 - ufp64_value_2;
+```
+
+Mathematical functions
+
+Exponential Function
+```
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp64/README.md [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+## Getting Started
+
+In order to use the `UFP64` type it must be added to the Forc.toml file and then imported into your Sway project. To add Sway-libs as a dependency to the Forc.toml in your project, please see the [README.md](../../../README.md).
+
+```rust
+use sway_libs::ufp64::UFP64;
+```
+
+Once imported, a `UFP64` type can be instantiated defining a new variable and calling the `from` function.
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp64/README.md]: Fixed

**Review by simonr0204 — APPROVED** [review]
LGTM
